In [30]:
# Imports and pip installations (if needed)
import pandas as pd # import for pandas library
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression # import for logistic regression model
from sklearn.preprocessing import LabelEncoder
from sklearn.svm import SVC # import for SVM model
from sklearn.neural_network import MLPClassifier # import for neural network model
from sklearn.neighbors import KNeighborsClassifier # import for KNN classifier model
from sklearn.model_selection import cross_val_score # import for cross-validation

# Part 1: Load the dataset

In [31]:
# Load the dataset (load remotely or locally)
df = pd.read_csv('iris.csv')

# Output the first 15 rows of the data
print(df.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(df.describe(include="all"), '\n') # for describing the numerical data
print(df['species'].describe(include="all"), '\n') # for describing the categorical data
print(df['species'].value_counts(normalize=True) * 100, '\n') # for describing the categorical data


    sepal_length  sepal_width  petal_length  petal_width species
0            5.1          3.5           1.4          0.2  setosa
1            4.9          3.0           1.4          0.2  setosa
2            4.7          3.2           1.3          0.2  setosa
3            4.6          3.1           1.5          0.2  setosa
4            5.0          3.6           1.4          0.2  setosa
5            5.4          3.9           1.7          0.4  setosa
6            4.6          3.4           1.4          0.3  setosa
7            5.0          3.4           1.5          0.2  setosa
8            4.4          2.9           1.4          0.2  setosa
9            4.9          3.1           1.5          0.1  setosa
10           5.4          3.7           1.5          0.2  setosa
11           4.8          3.4           1.6          0.2  setosa
12           4.8          3.0           1.4          0.1  setosa
13           4.3          3.0           1.1          0.1  setosa
14           5.8         

## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The dataset provided is the Iris dataset, which includes measurements of various iris flowers. This data has four main features: sepal length, sepal width, petal length, and petal width, each recorded in centimeters. These features serve as the input data to help identify the flower’s species. The target label, or the variable we aim to predict, is the species of each iris, with three possible classes: setosa, versicolor, and virginica. Each row in the dataset corresponds to a single flower, with the species column identifying the actual class. The mapping of the label values to classes is direct, where each species label in the species column corresponds to one of the three species categories (setosa, versicolor, or virginica). When viewing the first few rows, we see the actual class labels like "setosa" in the species column, making the data easy to interpret. When using the describe function, we get a summary of the numerical columns, including the mean and standard deviation, but these statistics do not apply to the species column as it is categorical. For this column, counts and frequencies are more relevant for understanding data distribution. This data setup provides a clear structure for training a model to classify iris species based on these measurements.

# Part 2: Split the dataset into train and test

In [32]:
# Take the dataset and split it into our features (X) and label (y)
X = df.drop('species', axis=1) # Features (X)
y = df['species'] # Label (y)

print(df.shape) # no of columns, no of rows

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.10, random_state = 1234)

(150, 5)


# Part 3: Logistic Regression

In [33]:
# i. Use sklearn to train a LogisticRegression model on the training set
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data_lr = X_test.iloc[5]  # Selecting one sample from the test set
probabilities_lr = log_reg.predict_proba([sample_data_lr]) # predicting 
label_encoder = LabelEncoder() # encoding the target variable (species)
y_encoded = label_encoder.fit_transform(y)

# displaying the probabilities with class names
class_names = label_encoder.inverse_transform([0, 1, 2])  # Map encoded labels back to original species names
probabilities_with_classes_lr = {class_name: float(prob) for class_name, prob in zip(class_names, probabilities_lr[0])}
print("Predicted probabilities for each class:", probabilities_with_classes_lr)

# iii. Report on the score for Logistic regression model, what does the score measure?
score_lr = log_reg.score(X_test, y_test)
print("Score of the logistic regression model", score_lr)

# iv. Extract the coefficents and intercepts for the boundary line(s)
coefficients = log_reg.coef_
intercepts = log_reg.intercept_
print(f"Coefficients of the boundary line {coefficients}; Intercepts of the boundary line {intercepts}")

Predicted probabilities for each class: {'setosa': 0.9685951189109234, 'versicolor': 0.03140477027569802, 'virginica': 1.1081337861689404e-07}
Score of the logistic regression model 1.0
Coefficients of the boundary line [[-0.39470253  0.95217611 -2.43950767 -1.04271311]
 [ 0.52888219 -0.33764985 -0.19379191 -0.92165204]
 [-0.13417967 -0.61452625  2.63329958  1.96436515]]; Intercepts of the boundary line [  9.39069212   2.13104267 -11.52173478]


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


# Report on the score for Logistic Regression:

The logistic regression model achieved a perfect accuracy score of 1.0, which means it correctly classified every sample in the test dataset. This score specifically measures accuracy, which is the fraction of predictions where the model’s predicted class matches the actual class. Using around 10% of the data for testing, this score indicates that all instances in this subset were accurately classified. However, a perfect score, especially on a small test set, should be interpreted with caution, as it may reflect overfitting or a chance result rather than robust performance. 
The model also outputs predicted probabilities for each class (setosa, versicolor, virginica), reflecting its confidence level for each class. For example, for one sample, the model assigned a probability of nearly 99.5% to the virginica class, indicating strong confidence in this prediction. Additionally, the model provides coefficients for each feature—sepal length, sepal width, petal length, and petal width—which show the influence of each feature on the prediction for each class. The intercepts further adjust the decision boundaries between classes. For instance, petal length and petal width heavily influence the classification of setosa, while other classes like versicolor and virginica are influenced differently. Overall, while the model performs perfectly on this test split, evaluating it on a larger, more varied dataset would provide a better assessment of its real-world accuracy.

# Part 4: Support Vector Machine

In [34]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svc = SVC(probability=True) # support vector classifier
svc.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data_svm = X_test.iloc[7]
probabilities_svm = svc.predict_proba([sample_data_svm])
probabilities_with_classes_svm = {class_name: float(prob) for class_name, prob in zip(class_names, probabilities_svm[0])}
print("Predicted probabilities for each class:", probabilities_with_classes_svm)

# iii. Report on the score for the SVM, what does the score measure?
score_svm = svc.score(X_test, y_test)
print("Score of the SVM model is:", score_svm)

Predicted probabilities for each class: {'setosa': 0.9773934770108563, 'versicolor': 0.014358235497907084, 'virginica': 0.008248287491236826}
Score of the SVM model is: 1.0


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


# Report on the score for Support Vector Machine:

The SVM model achieved an accuracy score of 1.0, indicating perfect classification on the test dataset. This score represents the model’s accuracy, which measures the proportion of correct predictions out of the total predictions made. In this case, an accuracy of 1.0 means that the SVM model correctly identified the class of every instance in the test subset. Accuracy is a common metric for classification tasks and is especially useful in balanced datasets like this one, where each class has similar representation. However, since the test set only includes around 10% of the data, the high accuracy might be specific to this subset and may not fully represent the model’s generalization ability on unseen data.

The model also provides predicted probabilities for each possible class, offering insights into the model’s confidence level. For one sample, the model output probabilities of approximately 1.1% for setosa, 77.7% for versicolor, and 21.2% for virginica, with the highest probability for versicolor. This breakdown shows that the model is highly confident in its prediction of versicolor for this sample. While the perfect score is promising, it’s recommended to validate the model on a larger and more diverse dataset to ensure its robustness and consistency in performance across different data splits.

# Part 5: Neural Network

In [35]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp = MLPClassifier(random_state=0, max_iter=2000)
mlp.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data_mlp = X_test.iloc[10]
probabilities_mlp = svc.predict_proba([sample_data_mlp])
probabilities_with_classes_mlp = {class_name: float(prob) for class_name, prob in zip(class_names, probabilities_mlp[0])}
print("Predicted probabilities for each class:", probabilities_with_classes_mlp)

# iii. Report on the score for the Neural Network, what does the score measure?
score_mlp = mlp.score(X_test, y_test)
print("Score of the neural network model is:", score_mlp)

# iv: Experiment with different options for the neural network, report on your best configuration
# Defining the configurations
configurations = [
    {'hidden_layer_sizes': [50, 50], 'activation': 'relu', 'solver': 'adam', 'alpha': 0.0001},
    {'hidden_layer_sizes': [100], 'activation': 'tanh', 'solver': 'adam', 'alpha': 0.0001},
    {'hidden_layer_sizes': [100, 50, 25], 'activation': 'relu', 'solver': 'lbfgs', 'alpha': 0.001},
    {'hidden_layer_sizes': [50], 'activation': 'logistic', 'solver': 'sgd', 'alpha': 0.01, 'learning_rate': 'adaptive'},
    {'hidden_layer_sizes': [100, 100], 'activation': 'relu', 'solver': 'adam', 'alpha': 0.0001},
]

best_score = 0
best_config = None
best_cv_score = 0
best_complexity = float('inf')  # Tracking model complexity as total neurons

for config in configurations:
    mlp = MLPClassifier(random_state=0, max_iter=2000, tol=1e-3, **config)
    mlp.fit(X_train, y_train)
    
    # Computing cross-validation score to assess generalization
    cv_score = cross_val_score(mlp, X_train, y_train, cv=5).mean()
    score = mlp.score(X_test, y_test)
    
    # Calculating model complexity as the total number of neurons
    complexity = sum(config['hidden_layer_sizes']) if isinstance(config['hidden_layer_sizes'], (list, tuple)) else config['hidden_layer_sizes']
    
    # Determining if this is the best configuration
    if score > best_score or (score == best_score and cv_score > best_cv_score) or (score == best_score and cv_score == best_cv_score and complexity < best_complexity):
        best_score = score
        best_cv_score = cv_score
        best_config = config
        best_complexity = complexity

    print(f"Configuration: {config}, Test Score: {score}, CV Score: {cv_score}, Complexity: {complexity}")

print(f"Best Configuration: {best_config}, Best Test Score: {best_score}, Best CV Score: {best_cv_score}")

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


Predicted probabilities for each class: {'setosa': 0.010739046194120095, 'versicolor': 0.9858766050160762, 'virginica': 0.0033843487898035885}
Score of the neural network model is: 1.0
Configuration: {'hidden_layer_sizes': [50, 50], 'activation': 'relu', 'solver': 'adam', 'alpha': 0.0001}, Test Score: 1.0, CV Score: 0.9703703703703702, Complexity: 100
Configuration: {'hidden_layer_sizes': [100], 'activation': 'tanh', 'solver': 'adam', 'alpha': 0.0001}, Test Score: 1.0, CV Score: 0.9777777777777779, Complexity: 100
Configuration: {'hidden_layer_sizes': [100, 50, 25], 'activation': 'relu', 'solver': 'lbfgs', 'alpha': 0.001}, Test Score: 1.0, CV Score: 0.9703703703703702, Complexity: 175
Configuration: {'hidden_layer_sizes': [50], 'activation': 'logistic', 'solver': 'sgd', 'alpha': 0.01, 'learning_rate': 'adaptive'}, Test Score: 0.2, CV Score: 0.3481481481481481, Complexity: 50
Configuration: {'hidden_layer_sizes': [100, 100], 'activation': 'relu', 'solver': 'adam', 'alpha': 0.0001}, Test

# Report on the score

The score for the neural network model is 1.0, indicating perfect accuracy on the test dataset, where around 10% of the data was used for testing. This score measures accuracy, which is the proportion of correctly classified samples out of the total samples in the test set. An accuracy of 1.0 means that the model correctly identified the class of every sample in this subset, demonstrating strong performance in this specific test instance. However, it's important to note that while a perfect score suggests that the model has effectively learned the patterns in the training data, this result might not fully represent the model’s generalization ability. Given the small test set, the model’s performance might vary with a larger or more diverse test set. 

# Report on the best neural network configuration

The best configuration for the neural network model is {'hidden_layer_sizes': [50, 50], 'activation': 'relu', 'solver': 'adam', 'alpha': 0.0001}, which achieved a perfect test accuracy score of 1.0 and a strong cross-validation (CV) score of 0.9704. This configuration consists of two hidden layers with 50 neurons each, providing a balanced level of complexity that allows the model to capture relationships within the data without excessive overfitting. The relu activation function is used, which is known to handle non-linear relationships effectively and is computationally efficient. The adam solver, a widely used optimization algorithm, enhances convergence, particularly for models with large datasets or higher complexity. The alpha parameter is set to 0.0001, which introduces a small degree of regularization to prevent overfitting without being overly restrictive.

The model’s performance with this configuration demonstrates an optimal generalization, as indicated by the cross-validation score of 0.9704, suggesting that the model performs consistently well across different data splits. Furthermore, the model complexity, calculated as the total number of neurons, is 100 for this configuration, which is relatively moderate compared to other high-performing configurations with higher complexity (e.g., [100, 100]). This supports the choice of [50, 50] as it balances high accuracy and generalization while maintaining a simpler, computationally efficient structure.

# Part 6: K-Nearest Neighbors

In [36]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
knn = KNeighborsClassifier(n_neighbors=7)
knn.fit(X_train, y_train)

# ii. For a sample datapoint, predict the probabilities for each possible class
sample_data_knn = X_test.iloc[10]
probabilities_knn = svc.predict_proba([sample_data_knn])
probabilities_with_classes_knn = {class_name: float(prob) for class_name, prob in zip(class_names, probabilities_knn[0])}
print("Predicted probabilities for each class:", probabilities_with_classes_knn)

# iii. Report on the score for kNN, what does the score measure?
score_knn = knn.score(X_test, y_test)
print("Score of the k-Neighbors Classifier score is:", score_knn)

Predicted probabilities for each class: {'setosa': 0.010739046194120095, 'versicolor': 0.9858766050160762, 'virginica': 0.0033843487898035885}
Score of the k-Neighbors Classifier score is: 1.0


/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  warnings.warn(


# Report on the score for KNN

The score for the k-Nearest Neighbors (kNN) classifier is 1.0, meaning the model achieved perfect accuracy on the test set. This score measures **accuracy**, which reflects the proportion of correctly classified instances out of the total number of instances in the test data. With an accuracy score of 1.0, the kNN model successfully identified the correct class for every sample in the test subset, indicating strong predictive performance for this data split.

In kNN, accuracy is calculated by comparing the predicted class of each test sample to its actual class label. The model bases its predictions on the majority class among the nearest neighbors, making kNN particularly sensitive to the distribution and density of classes in the data. While this perfect score suggests the model performed well on this test set, it’s important to consider that kNN's accuracy can vary with different test samples or dataset splits, especially in scenarios with complex decision boundaries. Testing on larger or more varied data and using cross-validation can provide a more comprehensive understanding of the model's generalization ability and robustness across different subsets.

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?

In this exercise, I evaluated multiple machine learning models—Logistic Regression, Support Vector Machine (SVM), Neural Network, and K-Neighbors Classifier—across different test data splits (10%, 20%, and 30%). With a testing size of 10% and 20%, each model consistently achieved an accuracy score of 1.0, indicating flawless classification on the test data. However, as the test split increased to 30%, I observed slight declines in accuracy for the Logistic Regression, SVM, and KNN models, each dropping to 0.9778, while the Neural Network maintained a perfect score. This suggests that the Neural Network, particularly the configuration with a hidden layer of 100 neurons, activation function "tanh," and solver "adam," demonstrated the most robustness across different splits, highlighting its strong generalization even with reduced training data.

One surprising observation was the stability of the Neural Network model, which retained a perfect score despite a larger test split. I believe this is due to the model’s complexity and adaptability with multiple hidden layer configurations, allowing it to capture patterns effectively. In contrast, the slight accuracy drop in the Logistic Regression, SVM, and KNN models with a 30% test split indicates minor sensitivity to reduced training data, likely due to their reliance on linear or neighborhood-based decision boundaries. This exercise showed me how model selection and data partitioning can impact performance, especially when training data is limited. Overall, I found that the Neural Network outperformed the other models in terms of stability and adaptability, making it the best choice for this dataset.